database.cc

```diff
@@ -1282,8 +1282,35 @@ std::ostream& operator<<(std::ostream& out, const database& db) {
     return out;
 }
 
-future<mutation> database::do_apply_counter_update(column_family& cf, const frozen_mutation& fm, schema_ptr m_schema,
-                                                   db::timeout_clock::time_point timeout,tracing::trace_state_ptr trace_state) {
+static bool has_deleted_counter_cell(const mutation& state) {
+    bool result = false;
+    for (auto&& cr : state.partition().clustered_rows()) {
+        cr.row().cells().for_each_cell_until([&] (auto id, const atomic_cell_or_collection& ac_o_c) {
+            const auto& cdef = state.schema()->column_at(column_kind::regular_column, id);
+            if (ac_o_c.as_atomic_cell(cdef).is_dead(gc_clock::now())) {
```

Comment:
Using the local clock here introduces nondeterminism... Depending on how long it takes for the process to reach this place, the cell will be considered deleted or not.

You said that "updating a dead counter cell has no effect". How does the updating code check whether the cell is dead or not?

Reply:
I'm no expert here but... counters have no TTL..? So I assumed that tombstone is a definitive tombstone (deliberately introduced, independent on time). If a node sees it = dead cell; if it doesn't = there's not much we can do.
If you're thinking about inter-node exchange of counter data (incl. tombstones) and synchronization thereof - well, that opens a lot of other questions...

Reply:
There is no such thing as a "tombstone independent of time". Each tombstone has a timestamp.

That's why the `is_dead(T)` can even make sense: it compares the tombstone's timestamp with `T`. If `T < tombstone_ts`, the cell is not dead, otherwise it is.

We don't want to compare the tombstone with `now`, where `now` is defined according to the local clock of one of the nodes (`gc_clock::now()`), because that doesn't make any sense! The node's clock might be a week into the future compared to the client's clock, and the update's timestamp might use the client's clock. So he wouldn't see the update in the CDC log even though according to his time, the counter is still alive, and will be alive for one more week.

Reply:
Sorry, I made a mistake in the second paragraph which resulted in my post not making any sense.

A tombstone cell with timestamp `tombstone_ts` is considered dead at time `T` iff `T >= tombstone_ts`.

Suppose that the update's timestamp is `U`, node A's local clock is `NOW = U + week`, and node B's local clock is `NOW = U + 1 day`. Suppose there is a tombstone `T = U + 6 days`.

If node B reads the counter, it will show a value (for 5 more days). But if node A reads the counter, it will show that it's dead.
Therefore, when we perform the update with timestamp `U`, we should still log it because `U < T`. There are (or may be) nodes that still consider the counter alive.